src/ImageStats/BasicStatsCalculator.tcc

```diff
@@ -41,17 +43,22 @@ BasicStats<T>::BasicStats()
       sumSq(0) {}
 
 template <typename T>
-BasicStatsCalculator<T>::BasicStatsCalculator(const std::vector<T>& data)
+BasicStatsCalculator<T>::BasicStatsCalculator(const T* data, size_t data_size)
     : _min_val(std::numeric_limits<T>::max()),
       _max_val(std::numeric_limits<T>::lowest()),
       _sum(0),
       _sum_squares(0),
       _num_pixels(0),
-      _data(data) {}
+      _data(data),
+      _data_size(data_size) {}
 
 template <typename T>
 void BasicStatsCalculator<T>::reduce(const size_t start, const size_t end) {
     size_t i;
+    if (start == end || end > _data_size) {
```

Comment:
If the data is empty, this function will be called with a start of 0 and an end of 0, which I would expect to be a valid no-op if for some reason the data is empty.

On the other hand, if the end exceeds the bounds of the array, that will cause an invalid read, so I'd say we need a stronger precaution than a debug message.

I would check only for the second condition, set the end to the data size if it's triggered, and maybe upgrade the debug message to a warning (with a more specific message).

Reply:
> I don't feel that strongly about correcting the end automatically; if this is triggered it means an internal logic error in the code which should be fixed.

I agree that we should never find this case during execution, but we should have an error in case of flawed logic and should be fixed.